docs/reference/logits_processors.md

```diff
@@ -0,0 +1,51 @@
+# Logits Processors
+
+TODO: Explanation of what logits processors do
+
+TODO: List of `OutlinesLogitsProcessor`
+
+TODO: Example of how to implement TemperatureLogitsProcessor
+
+TODO: using logits processors with models directly vs using with outlines
+
+TODO: Explanation of pipelines
+
+TODO: Link to log logits
+
+## Using Logits Processors in Outlines
+
+TODO Explanation
+
+```
+import outlines
+```
+
+
+## Chaining Logits Processors
+
+```
+import outlines
+import outlines.processors as processors
+
+model = outlines.models.llamacpp(
+    repo_id="M4-ai/TinyMistral-248M-v2-Instruct-GGUF",
+    filename="TinyMistral-248M-v2-Instruct.Q4_K_M.gguf"
+)
```

Comment:
This doesn't seem to work with llamacpp, but it does work with transformers:

```python
import outlines
import outlines.processors as processors
model = outlines.models.transformers(
    "openaccess-ai-collective/tiny-mistral",
)
# Create a chained logits processor
logits_processor = (
    processors.sequence_logging(model.tokenizer) |  # Log the generated sequence
	processors.logits_logging(model.tokenizer) |  # Log the raw logits
	processors.regex(r"[0-9]*", model.tokenizer) |  # Restrict the logits to match the pattern
	processors.temperature(0.5) |  # Set temperature to 0.5
	processors.logits_logging(model.tokenizer)  # Log the restricted, temperature-augmentent, sampled logits
)
generator = outlines.generate.base(model, logits_processor)
generator("What is your favorite number? ")
```

The error was

```python
Traceback (most recent call last):
  File "/home/cameron/dottxt/outlines/demo-logging.py", line 16, in <module>
    generator("What is your favorite number? ")
  File "/home/cameron/dottxt/outlines/outlines/generate/api.py", line 503, in __call__
    completions = self.model.generate(
  File "/home/cameron/dottxt/outlines/outlines/models/llamacpp.py", line 288, in generate
    completion = self.model(prompts, **llama_cpp_params)
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama.py", line 1799, in __call__
    return self.create_completion(
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama.py", line 1732, in create_completion
    completion: Completion = next(completion_or_chunks)  # type: ignore
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama.py", line 1216, in _create_completion
    for token in self.generate(
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama.py", line 810, in generate
    token = self.sample(
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama.py", line 704, in sample
    else logits_processor(self._input_ids[: idx + 1], logits)
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama.py", line 2250, in __call__
    scores = processor(input_ids, scores)
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/cameron/dottxt/outlines/outlines/processors/base_logits_processor.py", line 82, in __call__
    processed_logits = self.process_logits(
  File "/home/cameron/dottxt/outlines/outlines/processors/base_logits_processor.py", line 168, in process_logits
    result = processor.process_logits(input_ids, result)
  File "/home/cameron/dottxt/outlines/outlines/processors/logging.py", line 63, in process_logits
    self.logger.info(self.tokenizer.decode(input_ids))
  File "/home/cameron/dottxt/outlines/outlines/models/llamacpp.py", line 56, in decode
    decoded_bytes = self.tokenizer.detokenize(token_ids)
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/llama_tokenizer.py", line 52, in detokenize
    return self._model.detokenize(tokens)
  File "/home/cameron/dottxt/outlines/.venv/lib/python3.10/site-packages/llama_cpp/_internals.py", line 224, in detokenize
    self.model, llama_cpp.llama_token(token), buffer, size, 0, special
TypeError: 'list' object cannot be interpreted as an integer
```